File: 백준/Gold/34060. 로봇 청소기/로봇 청소기.py
# https://www.acmicpc.net/problem/34060
import sys
sys.setrecursionlimit(int(2e5))
input = sys.stdin.readline


def dfs(now):
    visited[now] = True
    for nxt in graph[now]:
        if not visited[nxt]:
            dfs(nxt)


n = int(input())
graph = [[] for _ in range(n)]
bef, now, last = dict(), dict(), -1
for idx in range(n):
    y = int(input())
    if y <= last:
        bef = now
        now = dict()
    elif y == last+1:
        graph[idx-1].append(idx)
        graph[idx].append(idx-1)

    last = y
    now[y] = idx

    if y in bef:
        bef_idx = bef[y]
        graph[bef_idx].append(idx)
        graph[idx].append(bef_idx)

count = 0
visited = [False] * n
for start in range(n):
    if not visited[start]:
        dfs(start)
        count += 1

print(f'{count}\n{n}')

File: 백준/Gold/34060. 로봇 청소기/로봇 청소기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;


public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());

        List<Integer>[] graph = new ArrayList[n];
        for (int i = 0; i < n; i++)
            graph[i] = new ArrayList<>();

        Map<Integer, Integer> bef = new HashMap<>();
        Map<Integer, Integer> now = new HashMap<>();
        int last = -1;

        for (int index = 0; index < n; index++) {
            int y = Integer.parseInt(br.readLine());

            if (y <= last) {
                bef = now;
                now = new HashMap<>();
            } else if (y == last+1) {
                graph[index].add(index-1);
                graph[index-1].add(index);
            }

            last = y;
            now.put(y, index);

            if (bef.containsKey(y)) {
                int befIndex = bef.get(y);
                graph[index].add(befIndex);
                graph[befIndex].add(index);
            }
        }

        int count = 0;
        boolean[] visited = new boolean[n];
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                bfs(graph, visited, i);
                count++;
            }
        }

        System.out.println(count + "\n" + n);
    }

    private static void bfs(List<Integer>[] graph, boolean[] visited, int start) {
        Queue<Integer> q = new ArrayDeque<>();
        q.add(start);
        visited[start] = true;

        while (!q.isEmpty()) {
            int now = q.poll();

            for (int nxt : graph[now]) {
                if (!visited[nxt]) {
                    visited[nxt] = true;
                    q.add(nxt);
                }
            }
        }
    }
}